pkg/services/ngalert/state/cache.go

```diff
@@ -137,72 +142,84 @@ func (c *cache) expandRuleLabelsAndAnnotations(ctx context.Context, alertRule *n
 	return expand(alertRule.Labels), expand(alertRule.Annotations)
 }
 
+func (c *cache) setAllStates(newStates map[int64]map[string]*ruleStates) {
+	c.mtxStates.Lock()
+	defer c.mtxStates.Unlock()
+	c.states = newStates
+}
+
 func (c *cache) set(entry *State) {
 	c.mtxStates.Lock()
 	defer c.mtxStates.Unlock()
 	if _, ok := c.states[entry.OrgID]; !ok {
-		c.states[entry.OrgID] = make(map[string]map[string]*State)
+		c.states[entry.OrgID] = make(map[string]*ruleStates)
 	}
 	if _, ok := c.states[entry.OrgID][entry.AlertRuleUID]; !ok {
-		c.states[entry.OrgID][entry.AlertRuleUID] = make(map[string]*State)
+		c.states[entry.OrgID][entry.AlertRuleUID] = &ruleStates{states: make(map[string]*State)}
 	}
-	c.states[entry.OrgID][entry.AlertRuleUID][entry.CacheId] = entry
+	c.states[entry.OrgID][entry.AlertRuleUID].states[entry.CacheId] = entry
 }
 
-func (c *cache) get(orgID int64, alertRuleUID, stateId string) (*State, error) {
+func (c *cache) get(orgID int64, alertRuleUID, stateId string) *State {
 	c.mtxStates.RLock()
 	defer c.mtxStates.RUnlock()
-	if state, ok := c.states[orgID][alertRuleUID][stateId]; ok {
-		return state, nil
+	ruleStates, ok := c.states[orgID][alertRuleUID]
+	if ok {
+		var state *State
+		state, ok = ruleStates.states[stateId]
+		if ok {
+			return state
+		}
 	}
-	return nil, fmt.Errorf("no entry for %s:%s was found", alertRuleUID, stateId)
+	return nil
 }
 
 func (c *cache) getAll(orgID int64) []*State {
 	var states []*State
 	c.mtxStates.RLock()
 	defer c.mtxStates.RUnlock()
 	for _, v1 := range c.states[orgID] {
-		for _, v2 := range v1 {
+		for _, v2 := range v1.states {
 			states = append(states, v2)
 		}
 	}
 	return states
 }
 
 func (c *cache) getStatesForRuleUID(orgID int64, alertRuleUID string) []*State {
-	var ruleStates []*State
+	var result []*State
 	c.mtxStates.RLock()
 	defer c.mtxStates.RUnlock()
-	for _, state := range c.states[orgID][alertRuleUID] {
-		ruleStates = append(ruleStates, state)
+	ruleStates, ok := c.states[orgID][alertRuleUID]
```

Comment:
Is it possible for this to cause a `nil` panic - could we ever have the orgID map uninitialized when this is called?

Reply:
technically, yes. In reality it does not happen. I will update anyway